README.md
# docker-squid

[![Build Status](https://travis-ci.com/madharjan/docker-squid.svg?branch=master)](https://travis-ci.com/madharjan/docker-squid)
[![Layers](https://images.microbadger.com/badges/image/madharjan/docker-squid.svg)](http://microbadger.com/images/madharjan/docker-squid)

Docker container for Squid Proxy based on [madharjan/docker-base](https://github.com/madharjan/docker-base/)

Squid configuration and Transparent Proxy configuration based on [jpetazzo/squid-in-a-can](https://github.com/jpetazzo/squid-in-a-can)

## Changes

* Squid and iptables in single container
* Install `iptables` config if run as Transparent Proxy and cleanup `iptables` config on container stop

## Features

* Environment variables to set upstream proxy & authentication
* Environment variables to set cache settings
* Bats ([sstephenson/bats](https://github.com/sstephenson/bats/)) based test cases

## Usages
* Run `docker-squid` as Transparent Proxy while `docker build` to speedup build time by caching  OS packages downloads
* Run `docker-squid` as Transparent Proxy while `docker build` behind Corporate Proxy WITHOUT changing Dockerfile *(works only if all downloads from Internet is HTTP only, NOT HTTPS)*
* Run `docker-squid` to conserve bandwidth on slow Internet connection by caching frequently downloaded files


## Squid 3.3.8 (docker-squid)

**Environment**

| Variable                  | Default | Example        |
|---------------------------|---------|----------------|
| DISABLE_SQUID             | 0       | 1 (to disable) |
| SQUID_INTERFACE_IP        | 0.0.0.0 | 170.17.42.1    |
| SQUID_HTTP_PORT           | 3128    | 8080           |
| SQUID_INTERCEPT_PORT      | 3129    | 8081           |
| SQUID_MAXIMUM_OBJECT_SIZE | 1024    | 512  (MB)      |
| SQUID_DISK_CACHE_SIZE     | 10000   | 1000           |
| SQUID_CACHE_PEER_HOST     |         | proxyHost      |
| SQUID_CACHE_PEER_PORT     |         | proxyPort      |
| SQUID_CACHE_PEER_AUTH     |         | user:pass      |
| ENABLE_TRANSPARENT_PROXY  | 0       | 1 (to enable)  |


## Build

### Clone this project

```bash
git clone https://github.com/madharjan/docker-squid
cd docker-squid
```

### Build Container

```bash
# login to DockerHub
docker login

# build
make

# tests
make run
make tests
make clean

# tag
make tag_latest

# release
make release
```

### Tag and Commit to Git

```bash
git tag 3.5.12
git push origin 3.5.12
```

## Run Container

### Prepare folder on host for container volumes
```
sudo mkdir -p /opt/docker/squid/cache/
sudo mkdir -p /opt/docker/squid/log/
```

### Run `docker-squid`

```bash
docker stop squid
docker rm squid

docker run -d \
  -p 8080:3128 \
  -e SQUID_CACHE_PEER_HOST=proxyHost \
  -e SQUID_CACHE_PEER_PORT=proxyPort \  
  -v /opt/docker/squid/cache:/var/cache/squid \
  -v /opt/docker/squid/log:/var/log/squid \
  --name squid \
  madharjan/docker-squid:3.5.12
```

## Run as Transparent Proxy

```bash
docker stop squid
docker rm squid

docker run -d \
  --network=host \
  --cap-add=NET_ADMIN \
  -e SQUID_HTTP_PORT=9090 \
  -e SQUID_INTERCEPT_PORT=9091 \
  -e SQUID_CACHE_PEER_HOST=proxyHost \
  -e SQUID_CACHE_PEER_PORT=proxyPort \  
  -e ENABLE_TRANSPARENT_PROXY=1 \
  -v /opt/docker/squid/cache:/var/cache/squid \
  -v /opt/docker/squid/log:/var/log/squid \
  --name squid \
  madharjan/docker-squid:3.5.12
```

## Run via Systemd

### Systemd Unit File - basic example

```txt
[Unit]
Description=Squid

After=docker.service

[Service]
TimeoutStartSec=0

ExecStartPre=-/bin/mkdir -p /opt/docker/squid/cache
ExecStartPre=-/bin/mkdir -p /opt/docker/squid/log
ExecStartPre=-/usr/bin/docker stop squid
ExecStartPre=-/usr/bin/docker rm squid
ExecStartPre=-/usr/bin/docker pull madharjan/docker-squid:3.5.12

ExecStart=/usr/bin/docker run \
  -p 8080:3128 \
  -v /opt/docker/squid/cache:/var/cache/squid \
  -v /opt/docker/squid/log:/var/log/squid \
  --name squid \
  madharjan/docker-squid:3.5.12

ExecStop=/usr/bin/docker stop -t 2 squid

[Install]
WantedBy=multi-user.target
```


Changelog.md
## 3.5.12

### Changes

* Updated to Ubuntu 16.04

## 3.3.8

### Changes

* Based on madharjan/docker-base:14.04
* Using Squid 3.3.8


test/run.sh
#!/bin/bash

set -x

./clean.sh

sudo mkdir -p /opt/docker/squid/cache
sudo mkdir -p /opt/docker/squid/log
sudo chown -R proxy:proxy /opt/docker/squid

docker run -d \
  --network=host \
  -e DEBUG=true \
  --cap-add=NET_ADMIN \
  -e SQUID_INTERFACE_IP=172.17.0.1 \
  -e SQUID_HTTP_PORT=9128 \
  -e SQUID_INTERCEPT_PORT=9129 \
  -e SQUID_DISK_CACHE_SIZE=15000 \
  -e ENABLE_TRANSPARENT_PROXY=1 \
  -v /opt/docker/squid/cache:/var/cache/squid3 \
  -v /opt/docker/squid/log:/var/log/squid3 \
  --name tsquid \
  madharjan/docker-squid:3.3.8

sleep 2
docker logs tsquid


test/clean.sh
#!/bin/bash

set -x

docker stop tsquid
docker rm tsquid
#sudo rm -rf /opt/docker/squid


services/12-squid.sh
#!/bin/bash

set -e

if [ "${DEBUG}" = true ]; then
  set -x
fi

DISABLE_SQUID=${DISABLE_SQUID:-0}
SQUID_INTERFACE_IP=${SQUID_INTERFACE_IP:-0.0.0.0}
SQUID_HTTP_PORT=${SQUID_HTTP_PORT:-3128}
SQUID_INTERCEPT_PORT=${SQUID_INTERCEPT_PORT:-3129}
SQUID_MAXIMUM_OBJECT_SIZE=${SQUID_MAXIMUM_OBJECT_SIZE:-1024}
SQUID_DISK_CACHE_SIZE=${SQUID_DISK_CACHE_SIZE:-10000}

SQUID_CACHE_PEER_HOST=${SQUID_CACHE_PEER_HOST:-}
SQUID_CACHE_PEER_PORT=${SQUID_CACHE_PEER_PORT:-}
SQUID_CACHE_PEER_AUTH=${SQUID_CACHE_PEER_AUTH:-}

ENABLE_TRANSPARENT_PROXY=${ENABLE_TRANSPARENT_PROXY:-0}

if [ ! "${DISABLE_SQUID}" -eq 0 ]; then
  touch /etc/service/squid/down
else
  rm -f /etc/service/squid/down
fi

sed -i "s/^#acl localnet/acl localnet/" /etc/squid/squid.conf
sed -i "s/^#http_access allow localnet/http_access allow localnet/" /etc/squid/squid.conf

sed -i "s/^#cache_dir .*/cache_dir ufs \/var\/cache\/squid ${SQUID_DISK_CACHE_SIZE} 16 256/" /etc/squid/squid.conf
sed -i "s/^# maximum_object_size .*/maximum_object_size ${SQUID_MAXIMUM_OBJECT_SIZE} MB/" /etc/squid/squid.conf

sed -i "s/^http_port .*/http_port ${SQUID_INTERFACE_IP}:${SQUID_HTTP_PORT}/" /etc/squid/squid.conf

echo "http_port ${SQUID_INTERFACE_IP}:${SQUID_INTERCEPT_PORT} intercept" >> /etc/squid/squid.conf

if [ ! x"${SQUID_CACHE_PEER_HOST}" = "x" ] && [ ! x"${SQUID_CACHE_PEER_PORT}" = "x" ]; then
  if [ ! x"${SQUID_CACHE_PEER_AUTH}" = "x" ]; then
    echo "cache_peer ${SQUID_CACHE_PEER_HOST} parent ${SQUID_CACHE_PEER_PORT} 0 proxy-only default no-query no-digest no-delay login=${SQUID_CACHE_PEER_AUTH}" >> /etc/squid/squid.conf
    echo "never_direct allow all" >> /etc/squid/squid.conf
  else
    echo "cache_peer ${SQUID_CACHE_PEER_HOST} parent ${SQUID_CACHE_PEER_PORT} 0 proxy-only default no-query no-digest no-delay" >> /etc/squid/squid.conf
    echo "never_direct allow all" >> /etc/squid/squid.conf
  fi
fi

cat <<EOT >> /etc/squid/squid.conf

# refresh pattern for debs and udebs
refresh_pattern deb$         20160 100%   20160
refresh_pattern udeb$        20160 100%   20160
refresh_pattern tar.gz$      20160 100%   20160
refresh_pattern Release$      1440  40%   20160
refresh_pattern Sources.gz$   1440  40%   20160
refresh_pattern Packages.gz$  1440  40%   20160
refresh_pattern cvd$          1440  40%   20160
refresh_pattern .                0  20%    1440
refresh_all_ims on
EOT

mkdir -p /var/cache/squid
chown -R proxy:proxy /var/cache/squid

/usr/sbin/squid -z

if [ "${ENABLE_TRANSPARENT_PROXY}" -eq 1 ]; then
  iptables -t nat -A PREROUTING -p tcp --dport 80 -j REDIRECT --to ${SQUID_INTERCEPT_PORT} -w
  iptables -t nat -L -n -v
fi


services/squid-startup.sh
#!/bin/bash

set -e

if [ "${DEBUG}" = true ]; then
  set -x
fi

DISABLE_SQUID=${DISABLE_SQUID:-0}
SQUID_INTERFACE_IP=${SQUID_INTERFACE_IP:-0.0.0.0}
SQUID_HTTP_PORT=${SQUID_HTTP_PORT:-3128}
SQUID_INTERCEPT_PORT=${SQUID_INTERCEPT_PORT:-3129}
SQUID_MAXIMUM_OBJECT_SIZE=${SQUID_MAXIMUM_OBJECT_SIZE:-1024}
SQUID_DISK_CACHE_SIZE=${SQUID_DISK_CACHE_SIZE:-10000}

SQUID_CACHE_PEER_HOST=${SQUID_CACHE_PEER_HOST:-}
SQUID_CACHE_PEER_PORT=${SQUID_CACHE_PEER_PORT:-}
SQUID_CACHE_PEER_AUTH=${SQUID_CACHE_PEER_AUTH:-}

ENABLE_TRANSPARENT_PROXY=${ENABLE_TRANSPARENT_PROXY:-0}

if [ ! "${DISABLE_SQUID}" -eq 0 ]; then
  touch /etc/service/squid/down
else
  rm -f /etc/service/squid/down
fi

sed -i "s/^#acl localnet/acl localnet/" /etc/squid3/squid.conf
sed -i "s/^#http_access allow localnet/http_access allow localnet/" /etc/squid3/squid.conf

sed -i "s/^#cache_dir .*/cache_dir ufs \/var\/cache\/squid3 ${SQUID_DISK_CACHE_SIZE} 16 256/" /etc/squid3/squid.conf
sed -i "s/^# maximum_object_size .*/maximum_object_size ${SQUID_MAXIMUM_OBJECT_SIZE} MB/" /etc/squid3/squid.conf

sed -i "s/^http_port .*/http_port ${SQUID_INTERFACE_IP}:${SQUID_HTTP_PORT}/" /etc/squid3/squid.conf

echo "http_port ${SQUID_INTERFACE_IP}:${SQUID_INTERCEPT_PORT} intercept" >> /etc/squid3/squid.conf

if [ ! x"${SQUID_CACHE_PEER_HOST}" = "x" ] && [ ! x"${SQUID_CACHE_PEER_PORT}" = "x" ]; then
  if [ ! x"${SQUID_CACHE_PEER_AUTH}" = "x" ]; then
    echo "cache_peer ${SQUID_CACHE_PEER_HOST} parent ${SQUID_CACHE_PEER_PORT} 0 proxy-only default no-query no-digest no-delay login=${SQUID_CACHE_PEER_AUTH}" >> /etc/squid3/squid.conf
    echo "never_direct allow all" >> /etc/squid3/squid.conf
  else
    echo "cache_peer ${SQUID_CACHE_PEER_HOST} parent ${SQUID_CACHE_PEER_PORT} 0 proxy-only default no-query no-digest no-delay" >> /etc/squid3/squid.conf
    echo "never_direct allow all" >> /etc/squid3/squid.conf
  fi
fi

cat <<EOT >> /etc/squid3/squid.conf

# refresh pattern for debs and udebs
refresh_pattern deb$         20160 100%   20160
refresh_pattern udeb$        20160 100%   20160
refresh_pattern tar.gz$      20160 100%   20160
refresh_pattern Release$      1440  40%   20160
refresh_pattern Sources.gz$   1440  40%   20160
refresh_pattern Packages.gz$  1440  40%   20160
refresh_pattern cvd$          1440  40%   20160
refresh_pattern .                0  20%    1440
refresh_all_ims on
EOT

mkdir -p /var/cache/squid3
chown -R proxy:proxy /var/cache/squid3

/usr/sbin/squid3 -z

if [ "${ENABLE_TRANSPARENT_PROXY}" -eq 1 ]; then
  iptables -t nat -A PREROUTING -p tcp --dport 80 -j REDIRECT --to ${SQUID_INTERCEPT_PORT} -w
  iptables -t nat -L -n -v
fi
